lib/src/main/kotlin/com/swmansion/starknet/provider/rpc/BuildJsonHttpDeserializer.kt

```diff
@@ -52,19 +69,37 @@ internal fun <T> buildJsonHttpDeserializer(
                 response.body,
             )
 
-        if (jsonRpcResponse.error != null) {
-            throw RpcRequestFailedException(
-                code = jsonRpcResponse.error.code,
-                message = jsonRpcResponse.error.message,
-                data = jsonRpcResponse.error.data,
+        extractResult(jsonRpcResponse, response)
+    }
+}
+
+internal fun <T> buildJsonBatchHttpDeserializer(
+    deserializationStrategies: List<KSerializer<T>>,
+    deserializationJson: Json,
+): HttpResponseDeserializer<List<T>> {
+    // TODO: In case of batch request, exception should not be thrown.
+    // Instead, we want to return wrapped responses.
+    // This enables access to successful results or encountered errors.
+
+    return Function { response ->
+        if (!response.isSuccessful) {
+            throw RequestFailedException(
                 payload = response.body,
             )
         }
 
-        if (jsonRpcResponse.result == null) {
-            throw RequestFailedException(message = "Response did not contain a result", payload = response.body)
+        val jsonArray = Json.parseToJsonElement(response.body).jsonArray
+        val jsonRpcResponses = deserializationStrategies.mapIndexed { index, strategy ->
+            deserializationJson.decodeFromString(
+                JsonRpcResponse.serializer(strategy),
+                jsonArray[index].toString(),
+            )
```

Comment:
1. Let's use `decodeFromJsonElement` instead to avoid repeat conversions
2. Mapping strategies to responses seems a little unintuitive to me, maybe we could map `jsonArray` instead?
3. I was wrong during our offline discussion. We do need to either access response `"id"` using raw json object or sort by it to ensure correct match between deserializer and response:
```suggestion
        val responses = jsonArray.map {
            deserializationJson.decodeFromJsonElement(
                JsonRpcResponse.serializer(deserializationStrategies[it.jsonObject["id"]!!.jsonPrimitive.int]), it,
            )
        }
```
Or since we don't support multiple types for now, we could go back to using a single deserializer.

Reply:
1. Fine
2. Agree, it may be more meaningful
3. Ok, good that we are both on the same page.

Regarding usage of single deserializer - it's actually more intuitive to pass deserializer for each response, plus incoming feature will require it. Do you agree we can keep list of deserializers?

Reply:
Yes, I marked it by accident, all the points should be addresses now.